pkg/pdp/replicator/replicate.go

```diff
@@ -0,0 +1,231 @@
+	// pull the data from the source
+	replicaResp, err := http.Get(task.Source.String())
+	if err != nil {
+		return fmt.Errorf("http get replication source (%s) failed: %w", task.Source, err)
+	}
+
+	// stream the source to the sink
+	req, err := http.NewRequest(http.MethodPut, task.Sink.String(), replicaResp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to create replication sink request: %w", err)
+	}
+	req.Header = replicaResp.Header
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed http PUT to replicate blob %s from %s to %s failed: %w", task.Blob.Digest, task.Source, task.Sink, err)
+	}
+	// verify status codes
+	if res.StatusCode >= 300 || res.StatusCode < 200 {
+		err := fmt.Errorf("unsuccessful http PUT to replicate blob %s from %s to %s status code %d", task.Blob.Digest, task.Source, task.Sink, res.StatusCode)
+		resData, err := io.ReadAll(res.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read replication sink response body: %w", err)
+		}
+		return fmt.Errorf("response body: %s: %w", resData, err)
+	}
+
+	// TODO this is a really gross way to have a dep, but can refactor later
+	acceptResp, err := storage.BlobAccept(ctx, r.storageService, &storage.BlobAcceptRequest{
+		Space: task.Space,
+		Blob:  task.Blob,
+		Put: blob.Promise{
+			UcanAwait: blob.Await{
+				Selector: ".out.ok",
+				// TODO IDK what this does, or what to put here, or if we even need it?
+				Link: task.Invocation.Link(),
+			},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to accept replication source blob %s: %w", task.Blob.Digest, err)
+	}
+
+	var forks []fx.Effect
+	forks = append(forks, fx.FromInvocation(acceptResp.Claim))
+
+	var pdpLink *ucan.Link
+	if acceptResp.Piece != nil {
+		// generate the invocation that will complete when aggregation is complete and the piece is accepted
+		pieceAccept, err := cap_pdp.Accept.Invoke(
+			r.storageService.ID(),
+			// TODO validate this is the correct audience
+			UploadServiceDID,
+			r.storageService.ID().DID().String(),
+			cap_pdp.AcceptCaveats{
+				Piece: *acceptResp.Piece,
+			}, delegation.WithNoExpiration())
+		if err != nil {
+			log.Errorw("creating piece accept invocation", "error", err)
+			return fmt.Errorf("creating piece accept invocation: %w", err)
+		}
+		pieceAcceptLink := pieceAccept.Link()
+		pdpLink = &pieceAcceptLink
+		forks = append(forks, fx.FromInvocation(pieceAccept))
+	}
+
+	ok := result.Ok[replica.TransferOK, ipld.Builder](replica.TransferOK{
+		Site: acceptResp.Claim.Link(),
+		PDP:  pdpLink,
+	})
+	rcpt, err := receipt.Issue(r.storageService.ID(), ok, ran.FromInvocation(task.Invocation))
+	if err != nil {
+		return fmt.Errorf("issuing receipt: %w", err)
+	}
+	if err := r.storageService.Receipts().Put(ctx, rcpt); err != nil {
+		return fmt.Errorf("failed to put transfer receipt: %w", err)
+	}
+	// TODO how does one send a receipt to the indexing service
+	client.Execute(rcpt, r.uploadServiceConn)
```

Comment:
@hannahhoward is there any prior art I can borrow from here? Looking to send a receipt for the transfer invocation to the upload service (sorry my TODO is wrong here)